### ISSUE TITLE
(yet another) inline function ambiguity 

### ISSUE THREAD

**Issue by StrayAlien** [dmn-tck/tck#203]
@etirelli @agilepro. Writing tests for "instance of".  Regardless of any runtime context this cannot be disambiguated without parenthesis.  (erm, as far as I can tell) ...

`
function() "foo" instance of function
`

My intention here is to test whether `function() "foo"` is a function.  As functions have no delimiters this parses as a function testing whether "foo" is a function.  Not as intended.

This will resolve it:

`
(function() "foo") instance of function
`

... as will imposing the limitation that func defs are not permitted in literal expressions (as per https://github.com/dmn-tck/tck/issues/197#issuecomment-443561951)

Without params, this would (likely) lead to something unintentional happening. 

Greg.

**Comment by etirelli** [comment]
Hi @StrayAlien 

You are moving fast. :) 

Just so we are on the same page, what is the goal of this ticket? To add tests for these cases?

Anyway, just to add some information, according to the precedence rules defined in the grammar, the following two syntaxes are equivalent (and will always return false when the function is invoked):

```
function() "foo" instance of function
function() ("foo" instance of function)
```

For the case you mention, where you want to test if a lambda function is of type function, the syntax you used is correct:
```
(function() "foo") instance of function
```

Further, DMN 1.3 will support additional type syntax for function, lists and contexts. For instance:
```
(function(bar:string) "foo") instance of function<string> -> string
```

> func defs are not permitted in literal expressions

Not sure I understand what do you mean by that? function definitions *are* permitted in literal expressions. Your example above is a valid case, as well as something like:
```
{ foo : function( x ) x + 1, bar : foo( 10 ) }.bar
```

The expression above returns 11 as a result.

**Comment by StrayAlien** [comment]
Hi there @etirelli.  Thanks.  Erm, I guess maybe I'm a but trigger happy re the issue, but essentially raising awareness I guess.  Happy to close.

Regarding:

> func defs are not permitted in literal expressions

That was not a statement of fact form the spec - but one of the 'possible solutions' I proposed with lambdas was to collectively agree *not* to inline them, then we can lazy-compile on execution - when we know exactly what is in scope.     

Thanks again.  Apols for late reply - I thought I had already replied, but, I have been living the life of a rockstar - head down writing tests at my desk .... :-)

Greg.
